### PR TITLE
setup_a_provider does not setup a new provider if a suitable one already exists

### PR DESCRIPTION
**Pull request by mfalesni**
Should be little bit more conserving our resources.

{{pytest: cfme/tests/automate/test_domain_priority.py cfme/tests/automate/test_namespace.py cfme/tests/configure/test_default_filters.py cfme/tests/configure/test_default_views.py cfme/tests/configure/test_access_control.py cfme/tests/configure/test_visual_cloud.py cfme/tests/configure/test_visual_infra.py cfme/tests/control/test_bugs.py cfme/tests/infrastructure/test_advanced_search_host.py cfme/tests/infrastructure/test_advanced_search_providers.py cfme/tests/infrastructure/test_advanced_search_vms.py cfme/tests/integration/test_openldap_auth.py cfme/tests/integration/test_aws_iam_auth_and_roles.py cfme/tests/integration/test_ldap_auth_and_roles.py cfme/tests/integration/test_ipa_external.py cfme/tests/intelligence/reports/test_canned_corresponds.py cfme/tests/intelligence/reports/test_report_chargeback.py cfme/tests/intelligence/reports/test_report_corresponds.py cfme/tests/intelligence/test_download_report.py cfme/tests/services/test_operations.py cfme/tests/test_rest.py cfme/tests/test_soap.py cfme/tests/cloud_infra_common/test_tag_objects.py -v --long-running }}

### CODE REVIEW

File: utils/providers.py [hunk 1]
```diff
@@ -147,6 +148,14 @@ def setup_a_provider(prov_class=None, prov_type=None, validate=True, check_exist
...
+    # If there is already a suitable provider, don't try to setup a new one.
+    already_existing = filter(is_provider_setup, providers)
+    if already_existing:
+        return setup_provider(  # This will run a refresh too
+            random.choice(already_existing), validate=validate, check_existing=check_existing)
+
+    # We need to setup a new one
     for provider in providers:
         try:
             result = setup_provider(provider, validate=validate, check_existing=check_existing)
```
Comment (psav) [inline utils/providers.py]: I _think_ this function ```setup_provider``` already does this check....but I could be wrong

Comment (psav) [inline utils/providers.py]: Oh but I see what you mean........If the _first_ one doesn't exist it will set it up.......

**Comment by dajoRH** [comment]
Lint report for commit 1087f593afcb3b25788032bcbe237c835287d6b7:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

File: utils/providers.py (hunk 1 above)
Comment (mfalesni) [inline utils/providers.py]: This checks whether any of the providers already exists and if it does, it does not setup any new one, just runs the validate.

**Comment by mfalesni** [comment]
@psav @seandst I am thinking whether we should introduce a bit of randomness inside the setup_a_provider function, because currently it seems to prefer certain providers because of the order. What if the provider set up would be determined by random choice from the eligible providers list rather than iterating the list? This might spread the load a bit throughout the providers. The code that I wrote in the PR would then secure that once a suitable provider is set up, it will be always reused.

**Comment by seandst** [comment]
I actually thought we already had that randomness, and I do think it's a good idea.

**Comment by mfalesni** [comment]
@seandst So I will play a little bit with that there.

**Comment by seandst** [comment]
Thanks!

**Comment by psav** [comment]
The only reason _I_ don't like randomness is because of repeatability. But.....who cares ;)

**Comment by psav** [comment]
The I in the previous comment is NOT A SLASH!

**Comment by mfalesni** [comment]
I added a `random.shuffle(providers)` there.

**Comment by dajoRH** [comment]
Lint report for commit 99485291303889608b812a9d5e07bca861ff5406:
:godmode: All seems good! :cake: :punch: :cookie: 
*CFME QE Bot*

**Comment by seandst** [comment]
> The only reason I don't like randomness is because of repeatability. But.....who cares ;)

My thinking is that if you're using `setup_a_provider`, the provider itself doesn't matter, and the randomness shouldn't affect the test. If the randomness does affect the test, it's a bug in our automation and we can thank the randomness for pointing it out. :)

In cases where the provider does matter, I think the existing fixtures and testgen work that we do is sufficient.

**Comment by mfalesni** [comment]
Despite the red colour, the tests appear to work properly with the providers, nothing seems breaking in this area.

**Comment by mfalesni** [comment]
@seandst ?

**Comment by seandst** [comment]
yup.

**Pull request merged**